### ISSUE TITLE
False positive 'possible null reference argument' after ensuring non-null via the null coalescing operator

### ISSUE THREAD

**Issue by Treit** [dotnet/roslyn#73426]
[Repro](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWsBuBDATogJYAmAzogLyIDiApgC4CS5AFAJSYyxEBmirVqTIB+AHQBhAPYBXAHYNEIkYnjsq1NbADesRPuQB2LgF8sMACJSAylIC2jABZE5AcyHlO5lADZkqRCtbBwZnN1YAGSIyBgAeFHgAPmJPHT0DFABODzJJWQUvGDNuOAg/KJj4iCSVemY2djSYAyNEORkAGw7TIA)

The following program results in a warning about passing a possible null argument to DoSomething. As far as I can see there is no possibility the argument could be null.

```
using System;
using System.Collections.Generic;

var ids = GetIds();

if ((ids?.Count ?? 0) == 0)
{
    return;
}

DoSomething(ids);

static void DoSomething(List<string> ids)
{
    Console.WriteLine(ids.Count);
}

static List<string>? GetIds()
{
    return null;
}
```

Replacing `if ((ids?.Count ?? 0) == 0)` with `if (ids is null || ids.Count == 0)` (or various other pattern matching techniques to accomplish the same thing) makes the warning go away, even though Sharplab says they lower to the same code.

We have this pattern in many places in a large codebase and it makes turning on nullable checks difficult. It would be much nicer if the compiler could detect that the variable cannot be null in this case.

**Comment by CyrusNajmabadi** [comment]
Closing out.  The compiler doesn't track the null relationship through arithmetic expressions (like `(ids?.Count ?? 0)` and then teh `==0`).

You can write this as: 

```c#
if (ids is null || ids.Count == 0)
{
    return;
}
```

or

```c#
if (ids is not { Count: > 0 })
    return;
```
